Reimbursement Requests Table

vilinh opened the pull request:
## Developer: Vi-Linh

Closes #21 

### Pull Request Summary
(Still uncomplete, but just putting up a PR for progress update)

Created reimbursement requests table component using data table component from shadcn. Added filtering for Organization, Payment Method, Status, and global filter search.

Uncomplete:  Filtering on date range (need date type), and collapsible rows for collapsible content.

### Modifications
- `src/components/reimbursement-table/collapsible-row.tsx`: Attempt to create collapsible row component wrapper
    - currently doesn't work because wrapper places entire row under one column instead of distributing cells (see screenshot)
- `src/components/reimbursement-table/column-def.tsx`: Column definition for reimbursement table.
- `src/components/reimbursement-table/fake-data.tsx`: Temporary mock data.
- `src/components/reimbursement-table/table.tsx`: Table component that includes the search filter and column filters.
- `src/components/ui/date-range-picker.tsx`: Date range picker component.
- `src/components/ui/table-column-filter-dropdown.tsx`: Dropdown component to do column filtering.
- `src/components/ui/debounced-input.tsx`: Debounced input for search input.
- `src/utils/format.ts`: Utility format functions to format transaction date and payment amount.

### Testing Considerations

Tested locally with interacting with dropdown filter and search filter.

### Pull Request Checklist

- [x] Code is neat, readable, and works
- [x] Comments are appropriate
- [x] The commit messages follows our [guidelines](https://h4i.notion.site/Conventional-Commits-593452ad1179489399ad3bd696ef772a)
- [x] The developer name is specified
- [x] The summary is completed
- [x] Assign reviewers

### Screenshots/Screencast

Base Table Functionality with Rows and Filtering
<img width="1381" alt="Screenshot 2024-02-03 at 9 41 01 AM" src="https://github.com/hack4impact-calpoly/ecologistics-portal/assets/81380688/3ef8f265-47d1-4cf7-90d8-1f70e0f240c6">

Attempt to Make Rows Expandable for Expanded Content
<img width="1380" alt="Screenshot 2024-02-03 at 9 39 55 AM" src="https://github.com/hack4impact-calpoly/ecologistics-portal/assets/81380688/71853e09-b2c1-47a3-b08c-2c3be5caf7cb">

ryanhu021 commented on src/components/reimbursement-table/collapsible-row.tsx:
```diff
@@ -0,0 +1,37 @@
+    <div>
+      <div
+        onClick={() => setIsOpen(!isOpen)}
+        style={{ cursor: "pointer", borderBottom: "1px solid #ccc" }}
```
if possible, could you try to use tailwind classes instead of raw CSS styles to keep things consistent?

ryanhu021 commented on src/components/reimbursement-table/column-def.tsx:
```diff
@@ -0,0 +1,79 @@
+import { Button } from "@/components/ui/button";
+import { DownloadIcon } from "@radix-ui/react-icons";
+
+export type Reimbursement = {
```
when the reimbursement schema PR gets merged, could we change all references to refer to the schema type instead of this for consistency? for now, this works great

ryanhu021 commented on src/components/ui/date-range-picker.tsx:
```diff
@@ -0,0 +1,65 @@
+"use client";
```
nit: the `components/ui` folder is generally used for shadcn components, so could we move custom UI components to a different folder (maybe `components/ui/custom`)?

ryanhu021 commented on src/components/ui/debounced-input.tsx:
```diff
@@ -0,0 +1,34 @@
+  debounce = 500,
+  ...props
+}: {
+  value: string | number;
```
to keep type definitions more organized, could we define prop types inside an interface?

ryanhu021 commented on src/utils/format.ts:
```diff
@@ -0,0 +1,16 @@
+  const day = date.getDate().toString().padStart(2, "0");
+  const year = date.getFullYear();
+
+  const formattedDate = `${month}/${day}/${year}`;
```
could this be done with `.toLocaleDateString()`?

ryanhu021 reviewed (commented):
this all looks great so far! this is definitely a large task so please take your time and don't feel like you have to get everything done asap :) also, don't worry about the dropdown for now since i believe the [newest mocks](https://www.figma.com/file/LjKZcnJrxFz8Sh2ou96EJz/Ecologistics-Non-Profit-Portal?type=design&node-id=540-1815&mode=design) have info popups instead of dropdowns. again, really great work and let me know if you need help with anything!

vilinh commented:
### Modifications

- edits according to review comments 
- `src/components/reimbursement-table/column-def.tsx`: added email to be able to pull from row
- `src/components/reimbursement-table/table.tsx`: added expandable section in the form of a row
  - hardcoded row for now (reimbursement schema does not have fields for what should be in expanded section)
  - since expandable section is made with a row component, will shift normal row content when expanded (tried many ways to try to wrap a separate component but either is inserted in one cell or results in a "Hydration error")
 
### Screenshots
Expanded:
![image](https://github.com/hack4impact-calpoly/ecologistics-portal/assets/81380688/574078be-6ad5-4716-95fc-5685d918d930)

Not Expanded:
![image](https://github.com/hack4impact-calpoly/ecologistics-portal/assets/81380688/97477fc6-dcef-470e-a66a-3fd508d1a43f)

vilinh commented:
### Modifications
- `src/components/reimbursement-table/column-def.tsx`: Add popover component for first cell

### Screenshots

Clicking on organization name:
<img width="544" alt="Screenshot 2024-02-18 at 2 27 16 PM" src="https://github.com/hack4impact-calpoly/ecologistics-portal/assets/81380688/ac2728bf-53ba-49b5-b208-a66636332e3c">
